web/packages/teleport/src/Discover/Shared/useJoinToken.ts

```diff
@@ -0,0 +1,101 @@
+import { useEffect, useState } from 'react';
+
+import { useTeleport } from 'teleport';
+
+import {
+  ResourceKind,
+  resourceKindToJoinRole,
+} from 'teleport/Discover/Shared/ResourceKind';
+
+import type { AgentLabel } from 'teleport/services/agents';
+import type { JoinMethod, JoinToken } from 'teleport/services/joinToken';
+
+interface SuspendResult {
+  promise?: Promise<any>;
+  response?: JoinToken;
+  error?: Error;
+}
+
+let abortController: AbortController;
+let joinTokenCache = new Map<ResourceKind, SuspendResult>();
+
+export function clearCachedJoinTokenResult(resourceKind: ResourceKind) {
+  joinTokenCache.delete(resourceKind);
+}
+
+export function useJoinToken(
+  resourceKind: ResourceKind,
+  suggestedAgentMatcherLabels: AgentLabel[] = [],
+  joinMethod: JoinMethod = 'token'
+): {
+  joinToken: JoinToken;
+  reloadJoinToken: () => void;
+} {
+  const ctx = useTeleport();
+
+  const [, rerender] = useState(0);
+
+  function run() {
+    abortController = new AbortController();
+
+    const result: SuspendResult = {
+      response: null,
+      error: null,
+      promise: ctx.joinTokenService
+        .fetchJoinToken(
+          {
+            roles: [resourceKindToJoinRole(resourceKind)],
+            method: joinMethod,
+            suggestedAgentMatcherLabels,
+          },
+          abortController.signal
+        )
+        .then(token => {
+          // Probably will never happen, but just in case, otherwise
+          // querying for the resource can return a false positive.
+          if (!token.internalResourceId) {
+            throw new Error(
+              'internal resource ID is required to discover the newly added resource, but none was provided'
+            );
+          }
+          result.response = token;
+        })
+        .catch(error => {
+          result.error = error;
+        }),
+    };
+
+    joinTokenCache.set(resourceKind, result);
+
+    return result;
+  }
+
+  useEffect(() => {
+    return () => {
+      abortController?.abort();
+    };
+  }, []);
+
+  const existing = joinTokenCache.get(resourceKind);
+
+  if (existing) {
+    if (existing.error) {
+      throw existing.error;
+    }
+
+    if (existing.response) {
+      return {
+        joinToken: existing.response,
+        reloadJoinToken() {
+          joinTokenCache.delete(resourceKind);
+
+          rerender(c => c + 1);
```

Comment:
```suggestion
          // Delete the cached token and force a rerender
          // so that this hook runs again and creates a new one.
          joinTokenCache.delete(resourceKind);
          rerender(c => c + 1);
```